### PR TITLE
select from mutation_fragments() + tablets: handle reads for non-owned partitions

### PR DESCRIPTION
**Pull request by denesb**
Attempting to read a partition via `SELECT * FROM MUTATION_FRAGMENTS()`, which the node doesn't own, from a table using tablets causes a crash.
This is because when using tablets, the replica side simply doesn't handle requests for un-owned tokens and this triggers a crash.
We should probably improve how this is handled (an exception is better than a crash), but this is outside the scope of this PR.
This PR fixes this and also adds a reproducer test.

Fixes: https://github.com/scylladb/scylladb/issues/18786

Fixes a regression introduced in 6.0, so needs backport to 6.0 and 6.1

### CODE REVIEW

**Comment by denesb** [comment]
> [background_reclaim](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10855/testReport/junit/boost/logalloc_test/background_reclaim) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+background_reclaim)

https://github.com/scylladb/scylladb/issues/15677

File: replica/mutation_dump.cc [hunk 1]
```diff
@@ -405,6 +405,20 @@ future<mutation_reader> make_partition_mutation_dump_reader(
...
+    // std::terminate because the replica side does not handle requests for
+    // un-owned tokens.
+    {
+        auto erm = tbl.get_effective_replication_map();
+        auto& topo = erm->get_topology();
+        const auto endpoints = erm->get_endpoints_for_reading(dk.token());
+        if (std::ranges::find(endpoints, topo.this_node()->endpoint()) == endpoints.end()) {
+            co_return make_empty_flat_reader_v2(output_schema, std::move(permit));
+        }
+    }
```
Comment (avikivity) [inline replica/mutation_dump.cc:420]: The check is only valid for the duration it is held. If we lose the range immediately after this }, we'll crash exactly the same.

We need to extend erm's lifetime over the whole query.

IMO the best approach is to accept it in make_partition_mutation_dump_reader() and pass it to mutation_dump_reader's constructor. This shows that we're locking the whole thing against tablet migrations.

Note: most likely, erm is in fact held for the duration of the query, but it's impossible to verify it.

Comment (denesb) [reply replica/mutation_dump.cc:420]: `erm` is held up in the `cql3` statement's `cql3::statements::mutation_fragments_select_statement::do_execute()`.

Comment (avikivity) [reply replica/mutation_dump.cc:420]: Then why not pass it down? So we can use local reasoning rather than non-local reasoning.

**Comment by denesb** [comment]
New in v2: enforce erm pinning by making it an argument of `dump_mutations()`.

File: replica/mutation_dump.cc (hunk 1 above)
Comment (denesb) [reply replica/mutation_dump.cc:420]: Done in v2.

**Review by avikivity — APPROVED** [review]

**Pull request merged**